source/models/row.py

```diff
@@ -0,0 +1,76 @@
+import os, json, uuid
+from outputs.exceptions import *
+from models.index import Index
+
+class Row:
+    def __init__(self, table, value = {}):
+        self.table = table
+        self.value = value
+        self.primary_key = self.__get_primary_key()
+
+    def __get_primary_key(self):
+        if self.table.table_metadata.primary_key not in self.value : 
+            self.value[self.table.table_metadata.primary_key] = str(uuid.uuid4().node)
+        else:
+            self.primary_key = self.value[self.table.table_metadata.primary_key]
+        return self.value[self.table.table_metadata.primary_key]
+
+    def serialize(self):
+        self.__lock()
+        self.__colomns_name_validate()
+        if self.table.table_metadata.primary_key not in self.value : 
+            self.value[self.table.table_metadata.primary_key] = str(uuid.uuid4().node)
+        row_json_data = json.dumps(self.value, indent = 2)
+        with open(self.get_path(), 'w') as row_file: 
+            row_file.write(row_json_data)
+        self.update_index()
+        self.__unlock()
+
+    def row_exists(self):
+        return os.path.isfile(self.get_path())
+
+    def get_path(self):
+        return os.path.join(self.table.get_path(), os.path.join("Data", self.primary_key + ".json"))
+
+    def __colomns_name_validate(self):
+        for row_colomn_name in self.value:
+            if row_colomn_name not in self.table.table_metadata.columns:
+                raise ColumnsNotExistInSchema(message = row_colomn_name + " is not exist in the schema of " + self.table.__table_name + " table")
+
+    def update_index(self):
+        for index_name in self.table.table_metadata.index_keys:
+            index = Index(self.table, index_name = index_name, index_value = self.value[index_name])
+            index.update_primary_key(primary_key = self.primary_key) 
+    
+    @staticmethod
+    def load_by_primary_key(table, primary_key):
+        row_file_path = os.path.join(table.get_path(), os.path.join("Data", primary_key + ".json"))
+        if not os.path.isfile(row_file_path) :
+            return None
+        with open(row_file_path, 'r') as row_file:
+            value = json.load(row_file)
+        return Row(table = table, value = value)
+
+    def delete(self):
+        self.__lock()
+        if not self.row_exists():
+            return
+        os.remove(self.get_path())
+        for index_name in self.table.table_metadata.index_keys:
+            index = Index(self.table, index_name = index_name, index_value = self.value[index_name])
```

Comment:
Metadata has a list of index object. You shouldn’t create Index in anywhere else.
Index shouldn’t have the value. It take the value in method index.add_primary_key, index.remove_primary_key

Reply:
ok now i get it, will work on it

Reply:
done